### PR TITLE
Allow different types of heading badges

### PR DESCRIPTION
**Pull request by piitaya**
## Proposed change

Allow different types of `heading badges` like we have for `badges`, `cards`, `rows`, etc...
Only `entity` type is allowed for now but this will open the possibility to have `area` item to show the temperature of the area for example or even custom `heading badges`.

Changes : 
- `entities` has be renamed to `badges` in the yaml file (with a migration).
- `string` item with an `entity_id` are not allowed anymore (breaking change but nobody use it)
- Added a error badge when type is wrong or entity_id doesn't exist

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/heading-items/types.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { Condition } from "../common/validate-condition";
+
+export type LovelaceHeadingItemConfig = {
+  type?: string;
+  [key: string]: any;
+  visibility?: Condition[];
+};
+
+export interface EntityHeadingItemConfig extends LovelaceHeadingItemConfig {
+  type?: "entity";
```
Comment (piitaya) [inline src/panels/lovelace/heading-items/types.ts:11]: The type is optional like we have for "row" but we can set it as mandatory.

File: src/panels/lovelace/cards/types.ts [hunk 2]
```diff
@@ -503,21 +504,12 @@ export interface TileCardConfig extends LovelaceCardConfig {
...
-
 export interface HeadingCardConfig extends LovelaceCardConfig {
   heading_style?: "title" | "subtitle";
   heading?: string;
   icon?: string;
   tap_action?: ActionConfig;
-  entities?: (string | HeadingEntityConfig)[];
+  items?: (LovelaceHeadingItemConfig | string)[];
+  /** @deprecated Use `triggers` instead */
+  entities?: (LovelaceHeadingItemConfig | string)[];
```
Comment (piitaya) [inline src/panels/lovelace/cards/types.ts:514]: Not sure we still need to support `string` here.

File: src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts [hunk 3]
```diff
@@ -56,7 +62,7 @@ export class HuiHeadingCardEditor
 
   public setConfig(config: HeadingCardConfig): void {
     assert(config, cardConfigStruct);
-    this._config = config;
+    this._config = migrateHeadingCardConfig(config);
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts:65]: If you do the migration before the assert we can remove entities from the struct

File: src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts [hunk 4]
```diff
@@ -175,22 +182,22 @@ export class HuiHeadingCardEditor
...
-  private _editEntity(ev: HASSDomEvent<{ index: number }>): void {
+  private _editBadge(ev: HASSDomEvent<{ index: number }>): void {
     ev.stopPropagation();
     const index = ev.detail.index;
-    const config = this._config!.entities![index!];
+    const config = this._badges(this._config!.badges)[index];
 
     fireEvent(this, "edit-sub-element", {
       config: config,
       saveConfig: (newConfig) => this._updateEntity(index, newConfig),
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts:192]:
```suggestion
      saveConfig: (newConfig) => this._updateBadge(index, newConfig),
```

File: src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts [hunk 5]
```diff
@@ -175,22 +182,22 @@ export class HuiHeadingCardEditor
...
-    } as EditSubElementEvent<HeadingEntityConfig>);
+      type: "heading-badge",
+    } as EditSubElementEvent<EntityHeadingBadgeConfig>);
   }
 
-  private _updateEntity(index: number, entity: HeadingEntityConfig) {
-    const entities = this._config!.entities!.concat();
-    entities[index] = entity;
-    const config = { ...this._config!, entities };
+  private _updateEntity(index: number, entity: EntityHeadingBadgeConfig) {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-heading-card-editor.ts:197]:
```suggestion
  private _updateBadge(index: number, entity: EntityHeadingBadgeConfig) {
```

File: src/panels/lovelace/editor/hui-entities-card-row-editor.ts [hunk 6]
```diff
@@ -14,7 +14,7 @@ import { EntityConfig, LovelaceRowConfig } from "../entity-rows/types";
 declare global {
   interface HASSDomEvents {
     "entities-changed": {
-      entities: LovelaceRowConfig[];
+      items: LovelaceRowConfig[];
```
Comment (bramkragten) [inline src/panels/lovelace/editor/hui-entities-card-row-editor.ts:17]: Same, still `entities-editor`?

Comment (piitaya) [reply src/panels/lovelace/editor/hui-entities-card-row-editor.ts:17]: Oops!

**Review by bramkragten — APPROVED** [review]

**Pull request merged**